// typedoc-theme-fresh/dist/index.js
import { JSX, DefaultTheme, DefaultThemeRenderContext, i18n, } from "typedoc";
// From src/lib/output/themes/lib.tsx
function classNames(names, extraCss) {
    const css = Object.keys(names)
        .filter((key) => names[key])
        .concat(extraCss || "")
        .join(" ")
        .trim()
        .replace(/\s+/g, " ");
    return css.length ? css : undefined;
}
// From src/lib/output/themes/lib.tsx
function wbr(str) {
    const parts = str.split(/(?<=[^A-Z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|(?<=[_-])(?=[^_-])/);
    return parts.flatMap((p) => [p, JSX.createElement("wbr", null)]).slice(0, -1);
}
// From src/lib/output/themes/default/partials/navigation.tsx
function buildSectionNavigation(context, headings) {
    const levels = [[]];
    function finalizeLevel(finishedHandlingHeadings) {
        const level = levels.pop();
        if (levels[levels.length - 1].length === 0 && finishedHandlingHeadings) {
            levels[levels.length - 1] = level;
            return;
        }
        const built = (JSX.createElement("ul", null, level.map((l) => (JSX.createElement("li", null, l)))));
        levels[levels.length - 1].push(built);
    }
    function getInferredHeadingLevel(heading) {
        if (heading.level) {
            // Regular heading
            return heading.level + 2;
        }
        if (heading.kind) {
            // Reflection
            return 2;
        }
        // Group/category
        return 1;
    }
    for (const heading of headings) {
        const inferredLevel = getInferredHeadingLevel(heading);
        while (inferredLevel < levels.length) {
            finalizeLevel(false);
        }
        while (inferredLevel > levels.length) {
            // Lower level than before
            levels.push([]);
        }
        levels[levels.length - 1].push(JSX.createElement("a", { href: heading.link, class: classNames({}, heading.classes) },
            heading.kind && context.icons[heading.kind](),
            JSX.createElement("span", null, wbr(heading.text))));
    }
    while (levels.length > 1) {
        finalizeLevel(true);
    }
    levels.unshift([]);
    finalizeLevel(true);
    return levels[0];
}
/**
 * Custom theme render context for the Fresh theme.
 * Extends the default TypeDoc render context with custom styling and layout.
 */
class FreshRenderContext extends DefaultThemeRenderContext {
    constructor(theme, page) {
        super(theme.router, theme, page, theme.application.options);
    }
    // Header: "tsd-page-title" class, breadcrumb and <h1>
    // override header = (props: PageEvent<Reflection>) => <>HEADER</>;
    // Remove footer, just the "Generated by TypeDoc" text
    footer = () => JSX.createElement(JSX.Fragment, null);
    // "Index" section on the page is redundant with the page navigation.
    index = () => JSX.createElement("div", { style: "margin-top: 64px" });
    // Override page navigation (right sidebar) to be a bit cleaner.
    pageNavigation = (props) => {
        if (!props.pageSections.some((sect) => sect.headings.length)) {
            return JSX.createElement(JSX.Fragment, null);
        }
        const sections = [];
        for (const section of props.pageSections) {
            if (section.title) {
                sections.push(JSX.createElement("details", { open: true, class: "tsd-accordion tsd-page-navigation-section" },
                    JSX.createElement("summary", { class: "tsd-accordion-summary", "data-key": `section-${section.title}` },
                        this.icons.chevronDown(),
                        section.title),
                    JSX.createElement("div", null, buildSectionNavigation(this, section.headings))));
            }
            else {
                sections.push(buildSectionNavigation(this, section.headings));
            }
        }
        return JSX.createElement("div", { class: "tsd-page-navigation" }, sections);
    };
    // Remove settings sidebar. "Member Visibility" filters are not needed, and
    // OS/light/dark theme toggle will be moved to header.
    settings = () => JSX.createElement(JSX.Fragment, null);
    toolbar = (props) => {
        return (JSX.createElement("header", { class: "tsd-page-toolbar" },
            JSX.createElement("div", { class: "tsd-toolbar-contents container" },
                JSX.createElement("a", { href: this.options.getValue("titleLink") ||
                        this.relativeURL("index.html"), class: "title" }, props.project.name),
                JSX.createElement("div", { id: "tsd-toolbar-links" }, Object.entries(this.options.getValue("navigationLinks")).map(([label, url]) => (JSX.createElement("a", { href: url }, label)))),
                JSX.createElement("div", { class: "tsd-theme-toggle md-show", style: "position: relative" },
                    JSX.createElement("div", { style: "position: absolute; left: 8px; top: 4px; pointer-events: none" },
                        JSX.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "20", height: "20", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "lucide lucide-sun-moon-icon lucide-sun-moon" },
                            JSX.createElement("path", { d: "M12 2v2" }),
                            JSX.createElement("path", { d: "M14.837 16.385a6 6 0 1 1-7.223-7.222c.624-.147.97.66.715 1.248a4 4 0 0 0 5.26 5.259c.589-.255 1.396.09 1.248.715" }),
                            JSX.createElement("path", { d: "M16 12a4 4 0 0 0-4-4" }),
                            JSX.createElement("path", { d: "m19 5-1.256 1.256" }),
                            JSX.createElement("path", { d: "M20 12h2" }))),
                    JSX.createElement("select", { id: "tsd-theme" },
                        JSX.createElement("option", { value: "os" }, "System"),
                        JSX.createElement("option", { value: "light" }, i18n.theme_light()),
                        JSX.createElement("option", { value: "dark" }, i18n.theme_dark()))),
                JSX.createElement("button", { id: "tsd-search-trigger", class: "tsd-widget", "aria-label": i18n.theme_search(), style: "\n              border: 1px solid var(--color-background-active);\n              border-radius: 4px;\n              margin-left: 8px;\n            " },
                    JSX.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "20", height: "20", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "lucide lucide-search-icon lucide-search", style: "margin-bottom: 1px; margin-left: 4px; margin-right: 4px;" },
                        JSX.createElement("path", { d: "m21 21-4.34-4.34" }),
                        JSX.createElement("circle", { cx: "11", cy: "11", r: "8" })),
                    JSX.createElement("span", { class: "lg-show", style: "margin-left: 4px; margin-right: 12ch;" }, "Search\u2026")),
                JSX.createElement("dialog", { id: "tsd-search", "aria-label": i18n.theme_search() },
                    JSX.createElement("input", { role: "combobox", id: "tsd-search-input", "aria-controls": "tsd-search-results", "aria-autocomplete": "list", "aria-expanded": "true", spellcheck: false, autocapitalize: "off", autocomplete: "off", placeholder: i18n.theme_search_placeholder(), maxLength: 100 }),
                    JSX.createElement("ul", { role: "listbox", id: "tsd-search-results" }),
                    JSX.createElement("div", { id: "tsd-search-status", "aria-live": "polite", "aria-atomic": "true" },
                        JSX.createElement("div", null, i18n.theme_preparing_search_index()))),
                JSX.createElement("a", { href: "#", class: "tsd-widget menu", id: "tsd-toolbar-menu-trigger", "data-toggle": "menu", "aria-label": i18n.theme_menu(), style: "\n              border: 1px solid var(--color-background-active);\n              border-radius: 4px;\n              margin-left: 4px;\n            " },
                    JSX.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "20", height: "20", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "lucide lucide-menu-icon lucide-menu", style: "margin-left: 4px; margin-right: 4px;" },
                        JSX.createElement("path", { d: "M4 5h16" }),
                        JSX.createElement("path", { d: "M4 12h16" }),
                        JSX.createElement("path", { d: "M4 19h16" }))))));
    };
}
// TODO: Maybe move this into a separate CSS file and set up a non-tsc bundler.
const styles = `
/* typedoc-theme-fresh */

/* @view-transition {
  navigation: auto;
} */

:root {
  --light-color-background: #fafafe;
  --light-color-background-active: #eaeced;
  --light-color-text: #222222;
  --light-color-text-aside: #5e5e5e;

  --dark-color-background: #1a1a1c;
  --dark-color-background-active: #2f2f2b;
  --dark-color-text: #dfdfd6;
  --dark-color-text-aside: #b0b0b0;

  --dim-header-height: 2.5rem;
  --dim-footer-height: 0px;

  --light-code-background: #ffffff18;
  --dark-code-background: #00000018;
}

body {
  padding: 0;
  margin: 0;
  & > .container.container-main {
    margin: 0 auto;
    margin-top: var(--dim-header-height);
    --dim-container-main-margin-y: 0px;
  }
}

@media (prefers-color-scheme: light) {
  ::selection {
    background-color: #11111a20;
  }
}

@media (prefers-color-scheme: dark) {
  ::selection {
    background-color: #bbbbff48;
  }
}

:root[data-theme="light"] {
  ::selection {
    background-color: #11111a20;
  }
}

:root[data-theme="dark"] {
  ::selection {
    background-color: #bbbbff48;
  }
}

h1, h2, h3, h4, h5, h6 {
  font-weight: 500;
}

.tsd-page-toolbar {
  position: fixed;
  background: linear-gradient(
    to bottom,
    color-mix(in srgb, var(--color-background) 80%, transparent) 60%,
    transparent
  );
  backdrop-filter: blur(2px);
  border-bottom: none;

  .title {
    font-weight: 500;
    font-size: 18px;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    min-width: 0;
    margin-right: auto;

    &:hover {
      color: var(--color-contrast-text);
      text-decoration: none;
    }
  }

  .tsd-theme-toggle {
    display: flex;
    position: relative;
    margin-right: 12px;
    color: var(--color-text-aside);
    
    select {
      cursor: pointer;
      border: 1px solid var(--color-background-active);
      border-radius: 4px;
      padding: 6px 8px;
      padding-left: 32px;  /* Space for the icon */
      font-size: 14px;
      background: transparent;
      transition: background-color 200ms;
      color: var(--color-text-aside);
      
      /* for Firefox and Chrome: these are inconsistent */
      -moz-appearance: none;
      -webkit-appearance: none;
      
      &:hover {
        background-color: var(--color-background-active);
      }
    }
  }
}

.tsd-widget {
  display: flex;
  align-items: center;
  height: auto;
  width: auto;
  padding: 4px;
  font-size: 14px;
  border-radius: 4px;
  transition: background-color 200ms;
  color: var(--color-text-aside);
  opacity: 100%;

  &:hover {
    background-color: var(--color-background-active);
  }
}

#tsd-search-results {
  & > li {
    padding-inline: 6px;
    transition: background-color 200ms;
  }

  a:hover {
    text-decoration: none;
  }
}

.col-content {
  overflow: hidden;
  padding-block: 24px;

  & > .tsd-typography {
    padding-bottom: 24px;
  }
}

.col-sidebar {
  top: var(--dim-header-height);  /* Relevant for tablet layout. */
}

.site-menu {
  top: var(--dim-header-height);
  padding-block: 24px;
  box-sizing: border-box;
}

.page-menu {
  top: var(--dim-header-height);
  padding-block: 48px 24px;
  box-sizing: border-box;
}

.tsd-navigation,
.tsd-page-navigation {
  > details {
    margin-bottom: 8px;
  }

  a {
    border-radius: 4px;
    padding-inline: 6px;
    margin-bottom: 2px;
  }
  a:hover {
    text-decoration: none;
    background-color: var(--color-background-active);
  }
  a.current {
    background-color: var(--color-background-active);
    span {
      color: var(--color-contrast-text);
    }
  }
}

.tsd-navigation {
  & > a {
    /* Top-level navigation anchor. */
    margin-block: 16px;
    font-size: 18px;
    background: transparent !important;
    padding: 0;
    
    &:hover {
      color: var(--color-contrast-text);
    }
  }
}

.tsd-page-navigation {
  border-left: 1px solid var(--color-background-active);
  padding-left: 4px;
  padding-block: 4px;
}

.tsd-page-navigation-section {
  & > summary > svg {
    margin-right: 0;
  }
  & > div {
    margin-left: 24px;
  }
}

.tsd-page-title {
  color: var(--color-contrast-text);
}

h1 > code.tsd-tag {
  margin-left: 12px;
}

code.tsd-tag {
  display: inline-block;
  transform: translateY(-0.1em);
  background: var(--color-background-secondary);
  border: 0;
  border-radius: 8px;
  padding: 2px 4px;
}

.tsd-breadcrumb {
  margin-bottom: 12px;
  li:after {
    content: " › ";
    opacity: 0.4;
  }
  li:last-of-type::after {
    content: "";
  }
}

.tsd-signature,
.tsd-signatures .tsd-signature {
  border: 0;
  border-radius: 8px;
  background: var(--color-background-secondary);
  font-weight: 500;
  padding: 10px 8px;
  line-height: 1.3;
}

.tsd-anchor-icon {
  color: var(--color-text-aside);
  padding-bottom: 1px;
}

.tsd-kind-icon {
  filter: grayscale(0.5) opacity(0.8);
}

/* Accordion icon */
.tsd-accordion-summary > svg {
  color: var(--color-text-aside);
  opacity: 50%;
  stroke-width: 1px;
  transition: transform 200ms ease-in-out;
}

.tsd-accordion-summary > svg:has(+ h2) {
  height: 24px;
  width: 24px;
}

/* Desktop sizing */
@media (min-width: 1200px) {
  .container-main {
    grid-template-columns: minmax(0, 1.1fr) minmax(0, 2.5fr) minmax(0, 17rem);
  }

  .site-menu {
    padding-right: 8px;
  }
  .page-menu {
    padding-left: 24px;
  }

  .tsd-widget.menu {
    display: none;
  }
}

/* Tablet sizing */
@media (min-width: 770px) and (max-width: 1199px) {
  .page-menu {
    padding-block: 24px 24px;
  }
  .site-menu {
    padding-block: 0 24px;
    margin-top: 0;
  }

  .tsd-widget.menu {
    display: none;
  }

  .lg-show {
    display: none !important;
  }
}

/* Mobile sizing */
@media (max-width: 769px) {
  .page-menu {
    display: none;
  }
  .site-menu {
    padding: 0;
  }

  .container {
    padding: 0 1rem;
  }

  .has-menu .col-sidebar {
    padding: 0 1rem;
    max-height: 100dvh;
  }

  .lg-show {
    display: none !important;
  }
  .md-show {
    display: none !important;
  }
}
`;
/**
 * Fresh theme for TypeDoc.
 *
 * A clean, minimalist and well-designed theme.
 */
export class Fresh extends DefaultTheme {
    constructor(renderer) {
        super(renderer);
        renderer.hooks.on("head.end", () => {
            // Insert some extra CSS at the end of the <head> element for this theme.
            return (JSX.createElement("style", { type: "text/css" },
                JSX.createElement(JSX.Raw, { html: styles })));
        });
        // Override some of the icons as needed.
        this.icons = {
            ...this.icons,
        };
    }
    getRenderContext(pageEvent) {
        return new FreshRenderContext(this, pageEvent);
    }
}
/**
 * Called by TypeDoc when loading this theme as a plugin.
 */
export function load(app) {
    app.renderer.defineTheme("fresh", Fresh);
}
